query-engine/core/src/query_graph_builder/write/delete.rs

```diff
@@ -20,37 +21,86 @@ pub(crate) fn delete_record(
     let where_arg = field.arguments.lookup(args::WHERE).unwrap();
     let filter = extract_unique_filter(where_arg.value.try_into()?, &model)?;
 
-    // Prefetch read query for the delete
-    let mut read_query = read::find_unique(field, model.clone(), query_schema)?;
-    read_query.add_filter(filter.clone());
-
-    let read_node = graph.create_node(Query::Read(read_query));
-    let delete_query = Query::Write(WriteQuery::DeleteRecord(DeleteRecord {
-        model: model.clone(),
-        record_filter: Some(filter.into()),
-    }));
-
-    let delete_node = graph.create_node(delete_query);
-    utils::insert_emulated_on_delete(graph, query_schema, &model, &read_node, &delete_node)?;
-
-    graph.create_edge(
-        &read_node,
-        &delete_node,
-        QueryGraphDependency::ProjectedDataDependency(
-            model.primary_identifier(),
-            Box::new(|delete_node, parent_ids| {
-                if !parent_ids.is_empty() {
-                    Ok(delete_node)
-                } else {
-                    Err(QueryGraphBuilderError::RecordNotFound(
-                        "Record to delete does not exist.".to_owned(),
-                    ))
-                }
+    if can_use_atomic_delete(query_schema, &model, &field) {
```

Comment:
I can't put a comment on a line that hasn't been changed but line 19, unless I'm missing something, should only be called when _not_ doing an atomic delete. Otherwise, we're losing the benefit of using return.

To be clear, I'm referring to `graph.flag_transactional()`. This indicates the core to open a transaction. There shouldn't be a need for a transaction if we're _just_ executing a delete.